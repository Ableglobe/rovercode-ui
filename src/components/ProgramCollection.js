import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import {
  Button,
  Card,
  Grid,
  Header,
  Input,
} from 'semantic-ui-react';

import CustomPagination from './CustomPagination';

class ProgramCollection extends PureComponent {
  handlePageChange = (e, { activePage }) => {
    const { onPageChange, owned } = this.props;

    onPageChange(activePage, owned);
  }

  handleSearchChange = (event) => {
    const { onSearchChange, owned } = this.props;

    onSearchChange(event.target.value, owned);
  }

  render() {
    const {
      label,
      onProgramClick,
      onRemoveClick,
      programs,
      owned,
    } = this.props;

    return (
      <Fragment>
        <Grid centered columns={3}>
          <Grid.Row>
            <Grid.Column />
            <Grid.Column>
              <Header as="h1" textAlign="center">
                {label}
              </Header>
            </Grid.Column>
            <Grid.Column textAlign="right">
              <Input
                className="prompt"
                icon="search"
                placeholder="Search..."
                onChange={this.handleSearchChange}
              />
            </Grid.Column>
          </Grid.Row>
        </Grid>
        <Card.Group centered>
          {
            programs.results.map(program => (
              <Card key={program.id}>
                <Card.Content>
                  <Card.Header>
                    {program.name}
                  </Card.Header>
                  <Card.Meta>
                    { owned ? 'Mine' : program.user.username }
                  </Card.Meta>
                </Card.Content>
                <Card.Content extra>
                  <Button primary id={program.id} onClick={onProgramClick}>
                    { owned ? 'Keep Working' : 'View' }
                  </Button>
                  {
                    owned ? (
                      <Button
                        negative
                        id={program.id}
                        name={program.name}
                        onClick={onRemoveClick}
                        floated="right"
                      >
                        Remove
                      </Button>
                    ) : (null)
                  }
                </Card.Content>
              </Card>
            ))
          }
        </Card.Group>
        {
          programs.total_pages > 1 ? (
            <Grid centered>
              <Grid.Row>
                <CustomPagination
                  defaultActivePage={1}
                  totalPages={programs.total_pages}
                  onPageChange={this.handlePageChange}
                />
              </Grid.Row>
            </Grid>
          ) : (null)
        }
      </Fragment>
    );
  }
}

ProgramCollection.defaultProps = {
  owned: false,
};

ProgramCollection.propTypes = {
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }).isRequired,
  label: PropTypes.string.isRequired,
  owned: PropTypes.bool,
  onProgramClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  onSearchChange: PropTypes.func.isRequired,
};

export default hot(module)(ProgramCollection);
