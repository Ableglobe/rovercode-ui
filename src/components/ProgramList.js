import React, { Component, Fragment } from 'react';
import {
  Button,
  Confirm,
  Icon,
  Loader,
  Segment,
} from 'semantic-ui-react';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

import ProgramCollection from './ProgramCollection';

const defaultState = {
  programLoaded: false,
  confirmOpen: false,
  focusProgram: {
    id: null,
    name: null,
  },
};

class ProgramList extends Component {
  constructor(props) {
    super(props);

    this.state = defaultState;
  }

  componentDidMount() {
    const { fetchPrograms, user } = this.props;
    return fetchPrograms({
      user: user.user_id,
    }).then(() => fetchPrograms({
      user__not: user.user_id,
    }));
  }

  showConfirm = e => this.setState({
    confirmOpen: true,
    focusProgram: {
      id: e.target.id,
      name: e.target.name,
    },
  })

  cancelRemove = () => this.setState(defaultState)

  removeProgram = () => {
    const { fetchPrograms, removeProgram, user } = this.props;
    const { focusProgram } = this.state;

    this.setState(defaultState);

    return removeProgram(focusProgram.id)
      .then(() => fetchPrograms({
        user: user.user_id,
      }))
      .then(() => fetchPrograms({
        user__not: user.user_id,
      }));
  }

  loadProgram = (e) => {
    const { changeReadOnly, fetchProgram } = this.props;
    const readOnly = e.target.dataset.owned === 'false';


    fetchProgram(e.target.id).then(() => {
      changeReadOnly(readOnly);
      this.setState({
        programLoaded: true,
      });
    });
  }

  fetch = (params, owned) => {
    const { fetchPrograms, user } = this.props;

    if (owned) {
      fetchPrograms({
        user: user.user_id,
        ...params,
      });
    } else {
      fetchPrograms({
        user__not: user.user_id,
        ...params,
      });
    }
  }

  programSegment = (programs, label, owned) => (
    <Segment raised style={{ margin: '10px 10% 10px 10%' }}>
      <ProgramCollection
        programs={programs}
        label={label}
        owned={owned}
        onProgramClick={this.loadProgram}
        onRemoveClick={this.showConfirm}
        onUpdate={this.fetch}
      />
    </Segment>
  )

  render() {
    const { programs, userPrograms } = this.props;
    const {
      confirmOpen,
      focusProgram,
      programLoaded,
    } = this.state;

    return (
      <Fragment>
        <Button primary as={Link} to="/mission-control" style={{ marginLeft: '10%' }}>
          <Icon name="plus" />
          New Program
        </Button>
        {
          programLoaded ? (
            <Redirect to={{
              pathname: '/mission-control',
            }}
            />
          ) : (null)
        }
        {
          userPrograms === null
            ? (<Loader active />)
            : this.programSegment(userPrograms, 'My Programs', true)
        }
        {
          programs === null
            ? (<Loader active />)
            : this.programSegment(programs, 'Find More', false)
        }
        <Confirm
          header="Remove Program"
          content={`Are you sure you want to remove ${focusProgram.name}?`}
          open={confirmOpen}
          onConfirm={this.removeProgram}
          onCancel={this.cancelRemove}
          cancelButton="No"
          confirmButton="Yes"
        />
      </Fragment>
    );
  }
}

ProgramList.defaultProps = {
  programs: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
  userPrograms: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
};

ProgramList.propTypes = {
  fetchProgram: PropTypes.func.isRequired,
  fetchPrograms: PropTypes.func.isRequired,
  removeProgram: PropTypes.func.isRequired,
  changeReadOnly: PropTypes.func.isRequired,
  user: PropTypes.shape({
    user_id: PropTypes.number.isRequired,
  }).isRequired,
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }),
  userPrograms: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }),
};

export default ProgramList;
