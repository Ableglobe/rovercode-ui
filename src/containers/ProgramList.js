import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { fetchProgram } from '../actions/code';
import { fetchPrograms, removeProgram } from '../actions/program';
import { updateValidAuth } from '../actions/auth';
import ProgramList from '../components/ProgramList';

const mapStateToProps = ({ program, user }) => ({ ...program, user });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  fetchProgram: (id) => {
    const fetchProgramAction = fetchProgram(id, {
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(fetchProgramAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
  fetchPrograms: (userIds, page) => {
    const xhrOptions = {
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    };

    if (userIds && userIds.include) {
      xhrOptions.params = {
        user: userIds.include,
      };
    }

    if (userIds && userIds.exclude) {
      xhrOptions.params = {
        user__not: userIds.exclude,
      };
    }

    if (page) {
      xhrOptions.params.page = page;
    }

    const fetchProgramsAction = fetchPrograms(xhrOptions);
    return dispatch(fetchProgramsAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
  removeProgram: (id) => {
    const removeProgramAction = removeProgram(id, {
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(removeProgramAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
});

const ProgramListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProgramList);

ProgramListContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(ProgramListContainer));
