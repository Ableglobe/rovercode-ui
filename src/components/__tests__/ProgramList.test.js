import React from 'react';
import { MemoryRouter } from 'react-router';
import { Loader } from 'semantic-ui-react';
import { Redirect } from 'react-router-dom';
import { mountWithIntl, shallowWithIntl } from 'enzyme-react-intl';
import ProgramCollection from '../ProgramCollection';
import ProgramList from '../ProgramList';

let changeReadOnly;
let fetchProgram;
let fetchPrograms;
let removeProgram;
let fetchTags;

describe('The ProgramList component', () => {
  beforeEach(() => {
    changeReadOnly = jest.fn();
    fetchProgram = jest.fn(() => Promise.resolve({}));
    fetchPrograms = jest.fn(() => Promise.resolve({}));
    removeProgram = jest.fn(() => Promise.resolve({}));
    fetchTags = jest.fn(() => Promise.resolve({}));
  });

  test('renders on the page with no errors', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();
    wrapper.setState({
      focusProgram: {
        id: 1,
        name: 'Test Program',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches programs on mount', async () => {
    await mountWithIntl(
      <MemoryRouter>
        <ProgramList
          changeReadOnly={changeReadOnly}
          fetchProgram={fetchProgram}
          fetchPrograms={fetchPrograms}
          removeProgram={removeProgram}
          fetchTags={fetchTags}
          user={{ user_id: 1 }}
        />
      </MemoryRouter>,
    );
    expect(fetchPrograms.mock.calls.length).toBe(2);
  });

  test('shows the correct number of programs for the user', async () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        userPrograms={programs}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(true);
    expect(wrapper.find(ProgramCollection).length).toBe(2);
    expect(wrapper.find(Loader).exists()).toBe(false);
  });

  test('shows loading when programs fetching', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
        programs={null}
      />,
    ).dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(true);
    expect(wrapper.find(ProgramCollection).length).toBe(1);
    expect(wrapper.find(Loader).exists()).toBe(true);
  });

  test('shows loading when user programs fetching', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
        userPrograms={null}
      />,
    ).dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(true);
    expect(wrapper.find(ProgramCollection).length).toBe(1);
    expect(wrapper.find(Loader).exists()).toBe(true);
  });

  test('redirects to mission control when program loads', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    wrapper.setState({
      programLoaded: true,
    });

    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).at(0).prop('to')).toEqual({
      pathname: '/mission-control',
    });
  });

  test('loads a program', async () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        programs={programs}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    await wrapper.instance().loadProgram({
      target: {
        parentNode: {
          id: undefined,
        },
        id: 33,
        dataset: {
          owned: 'false',
        },
      },
    });

    expect(changeReadOnly).toHaveBeenCalledWith(true);
    expect(fetchProgram).toHaveBeenCalledWith(33);
    expect(wrapper.state('programLoaded')).toBe(true);

    wrapper.setState({
      programLoaded: false,
    });

    await wrapper.instance().loadProgram({
      target: {
        parentNode: {
          id: 55,
          dataset: {
            owned: 'true',
          },
        },
      },
    });

    expect(changeReadOnly).toHaveBeenCalledWith(false);
    expect(fetchProgram).toHaveBeenCalledWith(33);
    expect(wrapper.state('programLoaded')).toBe(true);
  });

  test('fetches user programs after page change', async () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    await wrapper.instance().fetch({
      page: 2,
    }, true);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user: 1,
      page: 2,
    });
  });

  test('fetches other programs after page change', async () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    await wrapper.instance().fetch({
      page: 2,
    }, false);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user__not: 1,
      page: 2,
    });
  });

  test('fetches programs after search change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    wrapper.instance().fetch({
      search: 'abc',
      page: 1,
    }, true);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user: 1,
      page: 1,
      search: 'abc',
    });
  });

  test('fetches other programs after search change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    wrapper.instance().fetch({
      search: 'abc',
      page: 1,
    }, false);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user__not: 1,
      page: 1,
      search: 'abc',
    });
  });

  test('removes a program and reloads the program list', async () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    wrapper.setState({
      focusProgram: {
        id: 33,
        name: 'Unnamed_Design_3',
      },
    });
    await wrapper.instance().removeProgram();

    expect(fetchPrograms).toHaveBeenCalledTimes(4);
    expect(removeProgram).toHaveBeenCalledWith(33);
  });

  test('shows confirm dialog', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    wrapper.instance().showConfirm({
      target: {
        parentNode: {
          id: undefined,
        },
        id: 33,
        name: 'Unnamed_Design_3',
      },
    });

    expect(wrapper.state('confirmOpen')).toBe(true);
  });

  test('cancel dialog does not remove program', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        user={{ user_id: 1 }}
      />,
    ).dive();

    wrapper.instance().cancelRemove();

    expect(fetchPrograms).toHaveBeenCalledTimes(1);
    expect(removeProgram).not.toHaveBeenCalled();
    expect(wrapper.state('confirmOpen')).toBe(false);
  });
});
