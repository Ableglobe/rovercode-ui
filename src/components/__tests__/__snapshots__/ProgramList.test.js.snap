// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramList component renders on the page with no errors 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProgramList
    changeReadOnly={[MockFunction]}
    fetchProgram={[MockFunction]}
    fetchPrograms={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "user": 1,
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Promise {},
          },
        ],
      }
    }
    programs={
      Object {
        "next": null,
        "previous": null,
        "results": Array [],
        "total_pages": 1,
      }
    }
    removeProgram={[MockFunction]}
    user={
      Object {
        "user_id": 1,
      }
    }
    userPrograms={
      Object {
        "next": null,
        "previous": null,
        "results": Array [],
        "total_pages": 1,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Button
          as={[Function]}
          primary={true}
          style={
            Object {
              "marginLeft": "10%",
            }
          }
          to="/mission-control"
        >
          <Icon
            as="i"
            name="plus"
          />
          New Program
        </Button>,
        null,
        <Segment
          raised={true}
          style={
            Object {
              "margin": "10px 10% 10px 10%",
            }
          }
        >
          <ProgramCollection
            label="My Programs"
            onProgramClick={[Function]}
            onRemoveClick={[Function]}
            onUpdate={[Function]}
            owned={true}
            programs={
              Object {
                "next": null,
                "previous": null,
                "results": Array [],
                "total_pages": 1,
              }
            }
          />
        </Segment>,
        <Segment
          raised={true}
          style={
            Object {
              "margin": "10px 10% 10px 10%",
            }
          }
        >
          <ProgramCollection
            label="Find More"
            onProgramClick={[Function]}
            onRemoveClick={[Function]}
            onUpdate={[Function]}
            owned={false}
            programs={
              Object {
                "next": null,
                "previous": null,
                "results": Array [],
                "total_pages": 1,
              }
            }
          />
        </Segment>,
        <Confirm
          cancelButton="No"
          confirmButton="Yes"
          content="Are you sure you want to remove null?"
          header="Remove Program"
          onCancel={[Function]}
          onConfirm={[Function]}
          open={false}
          size="small"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": [Function],
          "children": Array [
            <Icon
              as="i"
              name="plus"
            />,
            "New Program",
          ],
          "primary": true,
          "style": Object {
            "marginLeft": "10%",
          },
          "to": "/mission-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "i",
              "name": "plus",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "New Program",
        ],
        "type": [Function],
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ProgramCollection
            label="My Programs"
            onProgramClick={[Function]}
            onRemoveClick={[Function]}
            onUpdate={[Function]}
            owned={true}
            programs={
              Object {
                "next": null,
                "previous": null,
                "results": Array [],
                "total_pages": 1,
              }
            }
          />,
          "raised": true,
          "style": Object {
            "margin": "10px 10% 10px 10%",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "label": "My Programs",
            "onProgramClick": [Function],
            "onRemoveClick": [Function],
            "onUpdate": [Function],
            "owned": true,
            "programs": Object {
              "next": null,
              "previous": null,
              "results": Array [],
              "total_pages": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ProgramCollection
            label="Find More"
            onProgramClick={[Function]}
            onRemoveClick={[Function]}
            onUpdate={[Function]}
            owned={false}
            programs={
              Object {
                "next": null,
                "previous": null,
                "results": Array [],
                "total_pages": 1,
              }
            }
          />,
          "raised": true,
          "style": Object {
            "margin": "10px 10% 10px 10%",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "label": "Find More",
            "onProgramClick": [Function],
            "onRemoveClick": [Function],
            "onUpdate": [Function],
            "owned": false,
            "programs": Object {
              "next": null,
              "previous": null,
              "results": Array [],
              "total_pages": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cancelButton": "No",
          "confirmButton": "Yes",
          "content": "Are you sure you want to remove null?",
          "header": "Remove Program",
          "onCancel": [Function],
          "onConfirm": [Function],
          "open": false,
          "size": "small",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Button
            as={[Function]}
            primary={true}
            style={
              Object {
                "marginLeft": "10%",
              }
            }
            to="/mission-control"
          >
            <Icon
              as="i"
              name="plus"
            />
            New Program
          </Button>,
          null,
          <Segment
            raised={true}
            style={
              Object {
                "margin": "10px 10% 10px 10%",
              }
            }
          >
            <ProgramCollection
              label="My Programs"
              onProgramClick={[Function]}
              onRemoveClick={[Function]}
              onUpdate={[Function]}
              owned={true}
              programs={
                Object {
                  "next": null,
                  "previous": null,
                  "results": Array [],
                  "total_pages": 1,
                }
              }
            />
          </Segment>,
          <Segment
            raised={true}
            style={
              Object {
                "margin": "10px 10% 10px 10%",
              }
            }
          >
            <ProgramCollection
              label="Find More"
              onProgramClick={[Function]}
              onRemoveClick={[Function]}
              onUpdate={[Function]}
              owned={false}
              programs={
                Object {
                  "next": null,
                  "previous": null,
                  "results": Array [],
                  "total_pages": 1,
                }
              }
            />
          </Segment>,
          <Confirm
            cancelButton="No"
            confirmButton="Yes"
            content="Are you sure you want to remove null?"
            header="Remove Program"
            onCancel={[Function]}
            onConfirm={[Function]}
            open={false}
            size="small"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": [Function],
            "children": Array [
              <Icon
                as="i"
                name="plus"
              />,
              "New Program",
            ],
            "primary": true,
            "style": Object {
              "marginLeft": "10%",
            },
            "to": "/mission-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "i",
                "name": "plus",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "New Program",
          ],
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ProgramCollection
              label="My Programs"
              onProgramClick={[Function]}
              onRemoveClick={[Function]}
              onUpdate={[Function]}
              owned={true}
              programs={
                Object {
                  "next": null,
                  "previous": null,
                  "results": Array [],
                  "total_pages": 1,
                }
              }
            />,
            "raised": true,
            "style": Object {
              "margin": "10px 10% 10px 10%",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "label": "My Programs",
              "onProgramClick": [Function],
              "onRemoveClick": [Function],
              "onUpdate": [Function],
              "owned": true,
              "programs": Object {
                "next": null,
                "previous": null,
                "results": Array [],
                "total_pages": 1,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ProgramCollection
              label="Find More"
              onProgramClick={[Function]}
              onRemoveClick={[Function]}
              onUpdate={[Function]}
              owned={false}
              programs={
                Object {
                  "next": null,
                  "previous": null,
                  "results": Array [],
                  "total_pages": 1,
                }
              }
            />,
            "raised": true,
            "style": Object {
              "margin": "10px 10% 10px 10%",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "label": "Find More",
              "onProgramClick": [Function],
              "onRemoveClick": [Function],
              "onUpdate": [Function],
              "owned": false,
              "programs": Object {
                "next": null,
                "previous": null,
                "results": Array [],
                "total_pages": 1,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cancelButton": "No",
            "confirmButton": "Yes",
            "content": "Are you sure you want to remove null?",
            "header": "Remove Program",
            "onCancel": [Function],
            "onConfirm": [Function],
            "open": false,
            "size": "small",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
